src/sign/ring.rs

```diff
@@ -351,6 +353,121 @@ impl fmt::Display for GenerateError {
 
 impl std::error::Error for GenerateError {}
 
+//------------ Nsec3HashError -------------------------------------------------
+
+/// An error when creating an NSEC3 hash.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum Nsec3HashError {
+    /// The requested algorithm for NSEC3 hashing is not supported.
+    UnsupportedAlgorithm,
+
+    /// Data could not be appended to a buffer.
+    ///
+    /// This could indicate an out of memory condition.
+    AppendError,
+
+    /// The hashing process produced an invalid owner hash.
+    ///
+    /// See: [OwnerHashError](crate::rdata::nsec3::OwnerHashError)
+    OwnerHashError,
+}
+
+/// Compute an [RFC 5155] NSEC3 hash using default settings.
+///
+/// See: [Nsec3param::default].
+///
+/// [RFC 5155]: https://www.rfc-editor.org/rfc/rfc5155
+pub fn nsec3_default_hash<N, HashOcts>(
+    owner: N,
+) -> Result<OwnerHash<HashOcts>, Nsec3HashError>
+where
+    N: ToName,
+    HashOcts: AsRef<[u8]> + EmptyBuilder + OctetsBuilder + Truncate,
+    for<'a> HashOcts: From<&'a [u8]>,
+{
+    let params = Nsec3param::<HashOcts>::default();
+    nsec3_hash(
+        owner,
+        params.hash_algorithm(),
+        params.iterations(),
+        params.salt(),
+    )
+}
+
+/// Compute an [RFC 5155] NSEC3 hash.
+///
+/// Computes an NSEC3 hash according to [RFC 5155] section 5:
+///
+/// > IH(salt, x, 0) = H(x || salt)
+/// > IH(salt, x, k) = H(IH(salt, x, k-1) || salt), if k > 0
+///
+/// Then the calculated hash of an owner name is:
+///
+/// > IH(salt, owner name, iterations),
+///
+/// Note that the `iterations` parameter is the number of _additional_
+/// iterations as defined in [RFC 5155] section 3.1.3.
+///
+/// [RFC 5155]: https://www.rfc-editor.org/rfc/rfc5155
+pub fn nsec3_hash<N, SaltOcts, HashOcts>(
+    owner: N,
+    algorithm: Nsec3HashAlg,
+    iterations: u16,
+    salt: &Nsec3Salt<SaltOcts>,
+) -> Result<OwnerHash<HashOcts>, Nsec3HashError>
+where
+    N: ToName,
+    SaltOcts: AsRef<[u8]>,
+    HashOcts: AsRef<[u8]> + EmptyBuilder + OctetsBuilder + Truncate,
+    for<'a> HashOcts: From<&'a [u8]>,
+{
+    if algorithm != Nsec3HashAlg::SHA1 {
+        return Err(Nsec3HashError::UnsupportedAlgorithm);
+    }
+
+    fn mk_hash<N, SaltOcts, HashOcts>(
+        owner: N,
+        iterations: u16,
+        salt: &Nsec3Salt<SaltOcts>,
+    ) -> Result<HashOcts, HashOcts::AppendError>
+    where
+        N: ToName,
+        SaltOcts: AsRef<[u8]>,
+        HashOcts: AsRef<[u8]> + EmptyBuilder + OctetsBuilder + Truncate,
+        for<'a> HashOcts: From<&'a [u8]>,
+    {
+        let mut buf = HashOcts::empty();
+
+        owner.compose_canonical(&mut buf)?;
+        buf.append_slice(salt.as_slice())?;
```

Comment:
I suspect this code was already so before I touched it, will have to check.

Reply:
`owner.compose_canonical()` needs a buffer to write into so that requires `buf` but the salt could be added to the context directly.

Reply:
Oh, right ... that's unfortunate.  We could replace the `N` parameter with `Name<Octs>` instead (which is likely what it will receive anyway), but I'll probably end up addressing that with the `new_name` module.